Match enumerable arguments by comparing contents rather than via Equals

blairconrad opened the pull request:
Fixes #1960.

blairconrad commented on src/FakeItEasy/Core/ArgumentEqualityComparer.cs:
```diff
@@ -36,5 +41,55 @@ public bool AreEqual(object expectedValue, object argumentValue)
+                var anyMoreExpected = expectedEnumerator.MoveNext();
+                var anyMoreActual = actualEnumerator.MoveNext();
+
+                while (anyMoreExpected && anyMoreActual)
```
I kind of miss [zip_longest](https://docs.python.org/3/library/itertools.html#itertools.zip_longest).

thomaslevesque commented on src/FakeItEasy/Core/ArgumentEqualityComparer.cs:
```diff
@@ -54,10 +62,33 @@ public bool AreEqual(object expectedValue, object argumentValue)
+                var anyMoreExpected = expectedEnumerator.MoveNext();
+                var anyMoreActual = actualEnumerator.MoveNext();
+
+                while (anyMoreExpected && anyMoreActual)
+                {
+                    if (!this.areEqualElements(expectedEnumerator.Current, actualEnumerator.Current))
+                    {
+                        return false;
+                    }
+
+                    anyMoreExpected = expectedEnumerator.MoveNext();
+                    anyMoreActual = actualEnumerator.MoveNext();
+                }
+
+                if (anyMoreExpected ^ anyMoreActual)
+                {
+                    return false;
+                }
```
Couldn't this be replaced by a call to the `SequenceEqual` overload that takes an equality comparer? of course we'd need an adapter from `IArgumentEqualityComparer` to `IEqualityComparer<T>`...

blairconrad replied:
Oh, yes. That's a much better idea.

blairconrad replied:
Doing now.

thomaslevesque reviewed (commented):
Thanks @blairconrad, nice work as usual!
I have a few suggestions

thomaslevesque commented on tests/FakeItEasy.Specs/CallMatchingSpecs.cs:
```diff
@@ -1014,6 +1014,38 @@ public static void SimpleEnumerableDoesNotMatchNonEnumerable(Action<object> fake
                 .x(() => A.CallTo(() => fake.Invoke(6)).MustNotHaveHappened());
         }
 
+        [Scenario]
+        public static void CharEnumerableMatchesString(Func<IEnumerable, int> fake, int result)
+        {
+            "Given a fake with a method that takes an enumerable parameter"
+                .x(() => fake = A.Fake<Func<IEnumerable, int>>());
+
+            "And I configure the method with an array of chars"
+                .x(() => A.CallTo(() => fake.Invoke(new[] { 'a', 'b', 'c' })).Returns(3));
+
+            "When I call the method with a string made of the same characters"
+                .x(() => result = fake.Invoke("abc"));
+
+            "Then it returns the configured value"
+                .x(() => result.Should().Be(3));
+        }
+
+        [Scenario]
+        public static void StringDoesNotMatchCharEnumerable(Func<IEnumerable, int> fake, int result)
+        {
+            "Given a fake with a method that takes an enumerable parameter"
+                .x(() => fake = A.Fake<Func<IEnumerable, int>>());
+
+            "And I configure the method with a string"
+                .x(() => A.CallTo(() => fake.Invoke("xyz")).Returns(-3));
+
+            "When I call the method with an enumerable of ofthe same characters"
+                .x(() => result = fake.Invoke(new List<char> { 'x', 'y', 'z' }));
+
+            "Then it returns the configured value"
+                .x(() => result.Should().Be(-3));
+        }
```
(In fact, we can't really check that the optimization was used, since the observed behavior is the same. We might need a unit test for that.)

blairconrad replied:
It doesn't check that we optimize, but does check that we get symmetric behaviour, which I had inadvertently broken when I had a different optimization. I will think on a unit test.

thomaslevesque replied:
Ah, I see. But it these two scenarios, the "optimized for string" path is not covered.

blairconrad replied:
It's true. I bet we have plenty of coverage for that case though. We must compare a couple of strings. I have no problem making explicit though, as a spec or unit test or both.

thomaslevesque replied:
True. Still, adding an explicit test would make it obvious this case wasn't forgotten.

blairconrad replied:
Specs for string matching and string mis-matching added.

blairconrad replied:
Integration test added to show that we actually use the optimization to compare strings.

blairconrad commented on src/FakeItEasy/Core/ArgumentEqualityComparer.cs:
```diff
@@ -47,11 +47,18 @@ private class EnumerableComparer : IArgumentEqualityComparer
-                    argumentEnumerable.Cast<object?>());
+            public bool AreEqual(object expectedValue, object argumentValue)
+            {
+                if (expectedValue is string expectedString && argumentValue is string argumentString)
+                {
+                    return string.Equals(expectedString, argumentValue);
+                }
```
My initial implementation did this, in fact. There's actually no need to change the priority in this case: you just insert it in the list of comparers where you want it, between the custom comparers and the one for the enumerable.

It gets a little tricky, though, if we we want to have the behaviour be symmetric (that is to work when the example object is a string and the actual is a non-string enumerable, and vice versa). My initial implementation didn't take these cases into account (so I added test cases that you comment on elsewhere) and opted for this implementation.

blairconrad commented on src/FakeItEasy/Core/ArgumentEqualityComparer.cs:
```diff
@@ -12,7 +13,10 @@ internal class ArgumentEqualityComparer
-            this.argumentEqualityComparers = argumentEqualityComparers.OrderByDescending(c => c.Priority).ToArray();
+            this.argumentEqualityComparers = argumentEqualityComparers
+                .OrderByDescending(c => c.Priority)
+                .Concat(new EnumerableComparer())
```
I can control the ordering if I register it here.
Also, I didn't think about it.
But now that I did, I still like being able to insert it after the ordered custom comparers.

blairconrad commented:
Thanks for the review, @thomaslevesque. Great points as always. I'll let your suggestions marinated, and you can enjoy my knee-jerk responses while I do! 😉

blairconrad commented:
More changes coming, @thomaslevesque. I just wanted to fixup the bad string comparison I had and also to restore `SequenceEqual` while I had the code.

thomaslevesque commented on src/FakeItEasy/Core/ArgumentEqualityComparer.cs:
```diff
@@ -57,39 +57,29 @@ public EnumerableComparer(Func<object?, object?, bool> elementComparer)
                 }
 
-                return true;
+                return Enumerable.SequenceEqual(
+                    ((IEnumerable)expectedValue).Cast<object?>(),
+                    argumentEnumerable.Cast<object?>(),
+                    this.elementComparer);
```
nit: why not use `SequenceEqual` as an extension method?

blairconrad replied:
I didn't think of it! Doing so now.

blairconrad commented:
Thanks again, @thomaslevesque. I think I addressed all your concerns, except for the ones for which you up-thumbed my rebuttals (directly inserting the `EnumerableComparer` into the list of available comparers and having the `EnumerableComparer` itself understand the optimization for comparing `string`s).

I'm happy to pursue changes to those, though, if I've misinterpreted your thumbs.
And as always, do tell me about any other concerns.

thomaslevesque commented on tests/FakeItEasy.IntegrationTests/CallMatchingTests.cs:
```diff
@@ -0,0 +1,30 @@
+            exception.Should().BeNull();
+        }
+
+        public class ThrowingCharComparer : ArgumentEqualityComparer<char>
+        {
+            protected override bool AreEqual(char expectedValue, char argumentValue)
+                => throw new InvalidOperationException("We don't want to compare chars individually");
+        }
```
That's a clever approach, I wouldn't have thought of doing it this way!

thomaslevesque reviewed (commented):
Thanks @blairconrad, this looks great, except for the F# imports.
No need to do another fixup commit for this, you can just squash when it's fixed.

blairconrad commented on tests/FakeItEasy.Specs/CallMatchingSpecs.cs:
```diff
@@ -10,8 +10,10 @@ namespace FakeItEasy.Specs
     using FakeItEasy.Tests.TestHelpers;
     using FakeItEasy.Tests.TestHelpers.FSharp;
     using FluentAssertions;
+    using Microsoft.FSharp.Data.UnitSystems.SI.UnitNames;
     using Xbehave;
     using Xunit;
+    using static Microsoft.FSharp.Core.ByRefKinds;
```
Man. No idea. I should pay more attention. Thanks for catching that.
(I'm surprised we get no build warnings for unused imports, tbh.)
I'll remove.

thomaslevesque reviewed (approved):
Thanks @blairconrad, nice work as usual!

blairconrad commented:
> Thanks @blairconrad, nice work as usual!

Eventually. Nice teamwork, I'd say.

Thanks again.
